### ISSUE TITLE
SC2154 string key misstaken for variable

### ISSUE THREAD

**Issue by pschoffer** [koalaman/shellcheck#1710]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or "online"): 0.4.6
- [/] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [/] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

numbers[one]=1
echo $((numbers[one] + 1))

```

#### Here's what shellcheck currently says:
```
Line 3:
numbers[one]=1
        ^-- SC2154: one is referenced but not assigned.
```

#### Here's what I wanted or expected to see:
No error, because `one` here is a string key not a variable. I tried putting it to single or double quotes, but no luck.

I guess at least `numbers['one']` should be here acceptable as that will not be considered variable by the arithmetic evaluation?  

Thanks for looking into this!

**Comment by beltaurus** [comment]
It looks like Bash itself is confused about this array:
When not declaring it as an (associative) array, we get:
```
$ numbers[one]=1
$ echo "${!numbers[@]}"
0
```
... the key of numbers is not the string, but an index?!

When trying to declare it as an associative array, we get a telling error message:
```
$ declare -A numbers
-bash: declare: numbers: cannot convert indexed to associative array
```
...indeed indicating that the numbers array is an INDEXED array...

So working around that:
```
$ unset numbers
$ declare -A numbers
$ numbers[one]=1
$ echo "${!numbers[@]}"
one
```
... the key is now correctly identified as the string; So let's try the final experiment:
```
#!/bin/bash
unset numbers
declare -A numbers
numbers[one]=1
echo "${!numbers[@]}"
echo $((numbers[one]+1))
```
...shellcheck now declares:
No issues detected!

Interesting, but since I always declare all my variables (just like I use 'strict' in Perl programming), I had not bumped into this!

**Comment by beltaurus** [comment]
I think it is related to how you initialize an associative array; The following script (without any declaration), also reports no issues in shellcheck:
```
#!/bin/bash
numbers=([one]=1)
echo "${!numbers[@]}"
echo $((numbers[one]+1))
```

**Comment by pschoffer** [comment]
Great stuff. Thank you very much for explaining. The workaround worked for me.

**Comment by koalaman** [comment]
ShellCheck is right: indices in arrays not declared associative are evaluated in an arithmetic context and not as strings. You can see that `one` *is* being treated as a variable by bash itself by setting it:

```
#!/bin/bash
one=42
numbers[one]=1
declare -p numbers
```

will output

```
declare -a numbers=([42]="1")
```

The fact that `numbers=([one]=1)` doesn't show a warning is an oversight in ShellCheck. It has the same problem. If you want `one` to be treated as a string, you have to `declare -A numbers` first.
